Mage.Sets/src/mage/cards/b/BraidsArisenNightmare.java

```diff
@@ -0,0 +1,148 @@
+    }
+
+    private BraidsArisenNightmareEffect(final BraidsArisenNightmareEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public BraidsArisenNightmareEffect copy() {
+        return new BraidsArisenNightmareEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+        TargetControlledPermanent target = new TargetControlledPermanent(1, 1, filter, true);
+        if (!target.canChoose(controller.getId(), source, game)) {
+            return false;
+        }
+        controller.chooseTarget(Outcome.Sacrifice, target, source, game);
+        Permanent permanent = game.getPermanent(target.getFirstTarget());
+        if (permanent == null) {
+            return false;
+        }
+        FilterControlledPermanent opponentFilter = new FilterControlledPermanent("a permanent that shares a card type with " + permanent.getName());
+        opponentFilter.add(new BraidsArisenNightmarePredicate(new HashSet<>(permanent.getCardType(game))));
+        if (!permanent.sacrifice(source, game)) {
+            return false;
+        }
+        for (UUID opponentId : game.getOpponents(controller.getId())) {
+            Player opponent = game.getPlayer(opponentId);
+            if (opponent == null) {
+                continue;
+            }
+            if (!braidsSacrifice(opponent, opponentFilter, game, source)) {
+                opponent.loseLife(2, game, source, false);
+                controller.drawCards(1, source, game);
+            }
+        }
+        return true;
+    }
+
+    private boolean braidsSacrifice(Player opponent, FilterControlledPermanent opponentFilter, Game game, Ability source) {
+        TargetControlledPermanent target = new TargetControlledPermanent(1, 1, opponentFilter, true);
+        if (!target.canChoose(opponent.getId(), source, game)) {
+            return false;
+        }
+        if (!opponent.chooseUse(Outcome.Sacrifice, "Sacrifice " + opponentFilter.getMessage() + '?', source, game)) {
+            return false;
+        }
+        opponent.chooseTarget(Outcome.Sacrifice, target, source, game);
+        Permanent permanent = game.getPermanent(target.getFirstTarget());
+        return permanent != null && permanent.sacrifice(source, game);
+    }
+}
+
+class BraidsArisenNightmarePredicate implements Predicate<MageObject> {
+
+    private final HashSet<CardType> cardTypes;
+
+    public BraidsArisenNightmarePredicate(HashSet<CardType> cardTypes) {
+        this.cardTypes = cardTypes;
+    }
+
+    @Override
+    public boolean apply(MageObject input, Game game) {
+        for (CardType type : input.getCardType(game)) {
+            if (cardTypes.contains(type)) {
+                return true;
+            }
+        }
+        return false;
+    }
+}
```

Comment:
This really should be a common predicate as there are ~20 cards which share a cards which use something like this: https://scryfall.com/search?q=o%3A%22shares+a+card+type%22&unique=cards&as=grid&order=released

Reply:
Yea, that's probably true.  I'll take a look at refactoring those.

Reply:
Most of those cards are not using filters/predicates.  I made a common predicate and refactored the ones that were in the latest commit.